# Problem 49 - Prime Permutations
(https://projecteuler.net/problem=49)

The sequence 1487, 4817, 8147 is an example of an
arithmetic sequence of primes that are permutations of one another.

There is one other such 3-element sequence of 4-digit primes. 

Find the concatenation of them in order.

In [50]:
# 3/11/19, on a whim
# S: 6:35
# E: 7:24

## Planning

possible approaches:
* brute force through all primes
    * brute force through all 4 digit primes
        * generate all 4 digit primes, generate all possible increasing combinations of 3, test them individually for (1) sequential increase, and (2) permutations
            * there are 1061 4-digit primes, so about 1 million combinations to check (less, accounting for increasing-ness) 
                * ~~also, no numbers can start wiith 0~~ 
        * ~~generate all 4 digit primes, for each one generate all possible digit permutations and test the resulting combinations for (1) primeness and (2) sequential increase~~

Functions:
* genPrimes(lower, upper) --> returns list of primes between bounds
* checkSequential(triple) --> returns if is an increasing sequence
* checkPermTriple(triple) --> returns if three numbers contain the same digits
* checkPerm(num1, num2) --> returns if two numbers contain the same digits
* getSequences() --> returns the two matching sequences

## Implementation


In [4]:
def genPrimes(lower, upper):
    """
    returns list of primes between bounds, 
    inclusive

    """
    def listElemDivides(list, dividend):
        """
        returns true if an element in list divides num
        """
        for divisor in list:
            if dividend % divisor==0:
                return True
        return False
    
    
    primes = [] # instantiation
    
    # generate all primes up to upper
    for num in range(2, upper+1):
        if not(listElemDivides(primes, num)):
            primes.append(num)
        
    # find index of lowest prime greater than lower 
    # linear search is fine
    lowerIdx = 0
    for i in range(len(primes)-1):
        if primes[i] <= lower and primes[i+1] >= lower:
            lowerIdx = i+1
    
    
    # return sublist of primes
    return primes[lowerIdx:]
    

In [5]:
# test
genPrimes(11,11) # perfected this with guess and check

[11]

In [6]:
# gen 4 digit primes
primes = genPrimes(1000, 9999)
print(primes)

[1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291, 1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451, 1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531, 1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601, 1607, 1609, 1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699, 1709, 1721, 1723, 1733, 1741, 1747, 1753, 1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889, 1901, 1907, 1913, 1931, 1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997, 1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083, 2087, 2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161, 2179, 2203, 2207, 2213, 2221, 2237, 2239, 2243, 225

In [7]:
# check
len(primes)
# nice

1061

In [8]:
def checkSequential(triple):
    """
    takes in a tuple of three numbers
    checks if they're in an increasing arithmetic sequence
    """
    x,y,z = triple
    return z-y == y-x

In [9]:
# test
triple1 = (1487, 4817, 8147)
print(checkSequential(triple1))

True


In [10]:
def checkPerm(num1, num2):
    """
    num1, num2 are two four-digit numbers
    checks if num1, num2 have the same digits
    """
    # assigns a prime number to each digit; if 
        # the products for the nums are equal, they're permutations
    prod1 = 1
    prod2 = 1
    
    def digitToPrime(d):
        """
        assigns and returns a prime for each digit
        """
        if d == 0:
            return 2
        elif d==1:
            return 3
        elif d==2:
            return 5
        elif d==3:
            return 7
        elif d==4:
            return 11
        elif d==5:
            return 13
        elif d==6:
            return 17
        elif d==7:
            return 19
        elif d==8:
            return 23
        elif d==9:
            return 29
        else:
            return 0 # for error catchign?
        
    while num1>0:
        prod1*=digitToPrime(num1%10)
        num1 //= 10
    
    while num2>0:
        prod2*=digitToPrime(num2%10)
        num2 //= 10
        
    return prod1==prod2

In [11]:
# testing
print(checkPerm(1487, 4817))
print(checkPerm(1487,1111))

True
False


In [12]:
def checkPermTriple(triple):
    x,y,z=triple # unpacking
    return checkPerm(x,y) and checkPerm(y,z)

In [13]:
# testing
print(checkPermTriple(triple1))
print(checkPermTriple( (11, 111, 1111) ))

True
False


In [24]:
def getSequences():
    """
    returns solution sequences in the form of a list
    """
    soln = []

    
    # generate primes
    primes = genPrimes(1000, 9999)
    length = len(primes)
    
    # code for generating all icnreasing prime triples
#     tripleList = []
#     for i in range(len(primes)):
#         for j in range(i+1, len(primes)):
#             for k in range(j+1, len(primes)):
#                 tripleList.append( (primes[i], primes[j], primes[k]) )
    
    # go through all three tuples (without storign), check them one by one
    for i in range(length):
        for j in range(i+1, length):
            for k in range(j+1, length):
                tempTriple = (primes[i], primes[j], primes[k])
                if checkSequential(tempTriple):
                    if checkPermTriple(tempTriple):
                        soln.append(tempTriple)

    #v2
#     for i in range(length):
#         for p2 in primes[i+1:]:
#             for p3 in primes[i+2:]:
#                 tempTriple = (primes[i], p2, p3)
#                 if checkSequential(tempTriple):
#                     if checkPermTriple(tempTriple):
#                         soln.append(tempTriple)
                        
        # check sequentialness first, it'll be faster than checking permutation-ness
        
    return soln

## Results

In [18]:
import timeit

In [19]:
start = timeit.default_timer()

solutions = getSequences()

end = timeit.default_timer()

print(f"Time taken: {end - start}")
# print(solutions)

Time taken: 71.42923610636149


In [22]:
print(solutions)

[(1487, 4817, 8147), (2969, 6299, 9629)]


In [25]:
start = timeit.default_timer()

solutions = getSequences()

end = timeit.default_timer()

print(f"Time taken: {end - start}")
# print(solutions)

Time taken: 123.4059062687611


In [ ]:
print(solutions)

### Answer:
The other triple of arithmetically increasing prime permutations is: (2969, 6299, 9629)

Their permutation is 296962999629

# Reflection

There are guys on the discussion forum who took less than 10 milliseconds to complete what took me 81 seconds. That's a difference of about 4 orders of magnitude....


In [52]:
81/0.007

11571.42857142857

Some guy took 1.5 seconds while using python, so it's not just a language speed thing...

In [68]:
import string
from primesieve import primes  #own written function
def anagrams(s):
    if s == "":
        return [s]
    else:
        ans = []
        for an in anagrams(s[1:]):
            for pos in range(len(an)+1):
                ans.append(an[:pos]+s[0]+an[pos:])
        return ans

p=primes(10000)

def check(num):
    global p
    per=anagrams(str(num))
    per=[int(i) for i in per]
    i=0
    while i<len(per):
        if per[i] not in p:
            del per[i]
        else:
            i+=1
    incr=[]
    for i in per:
        for j in per:
            if i-j>0:
                incr.append(i-j)
    for inc in incr:
        for prime in per:
            if prime+inc in p and prime+2*inc in p:
                return [prime, prime+inc,prime+inc*2]		
    return []

for prime in p:
    if prime not in check(1487) and prime>1000:
        a=check(prime)
        if len(a)==3:
            if a[1]-a[0]==3330:
                print a
                print str(a[0])+str(a[1])+str(a[2])
                break

SyntaxError: Missing parentheses in call to 'print'. Did you mean print(a)? (<ipython-input-68-b25da29cd66e>, line 41)

### Borrowed code from user "inhahe"

Resources used to update, replace primes module: 
* https://stackoverflow.com/questions/13326673/is-there-a-python-library-to-list-primes 
* http://openbookproject.net/thinkcs/python/english3e/classes_and_objects_I.html

In [70]:
class IterPrimes(object):
    def __init__(self,n=1):
        self.n=n

    def __iter__(self):
        return self

    def next(self):
        n = self.n
        while not sympy.isprime(n):
            n += 1
        self.n = n+1
        return n
    
def primes():
    p = IterPrimes()
    n = 2
    while True:
        yield n
        n = IterPrimes.next(p)
    

start = timeit.default_timer()    
    
p1= {} 

for prime in primes():
    if prime>=10000: break
    if prime>=1000:
        s=tuple(sorted(str(prime)))
        p1[s]=p1.get(s,[])+[prime]
p2=[sorted(v) for v in p1.values() if len(v)>=3]
for p in p2:
    for i,v in enumerate(p[:-2]):
        for i2, v2 in enumerate(p[i+1:-1]):
            for i3,v3 in enumerate(p[i2+1:]):
                if v3-v2==v2-v:
                    print (v, v2, v3)
                    
end = timeit.default_timer()
print(f"Time taken: {end-start}")



1487 4817 8147
2969 6299 9629
Time taken: 0.029823575477848863


# HOW???!!??!?!!?